### ISSUE TITLE
Rationalize PublishEvent* APIs

### ISSUE THREAD

**Issue by ItalyPaleAle** [dapr/go-sdk#174]
Currently, the Go SDK has 2 methods to publish events:

- [`PublishEvent`](https://github.com/dapr/go-sdk/blob/aec809a7d5dfda7c9d5e45d81b10db17481cb1f6/client/pubsub.go#L12)
- [`PublishEventfromCustomContent`](https://github.com/dapr/go-sdk/blob/aec809a7d5dfda7c9d5e45d81b10db17481cb1f6/client/pubsub.go#L35) (also note the F should have been uppercased here)

The latter doesn't really do much more than the former, besides calling `json.Marshal` internally and setting the ContentType to `application/json`.

With #173, I've proposed adding 2 more variants for those methods: `PublishEventWithMetadata` and `PublishEventfromCustomContentWithMetadata` (F is lowercased here for consistency). These address #164 and also allow doing things like setting the TTL.

If #173 is merged (which I hope it is!) we will have 4 methods.

Yet, there will still be things that can't be done: setting options that are already available in the protobuf definition, but can't be set using the higher-level PublishEvent* APIs. One example includes setting the `ContentType` of the document (it's set to `text/plain` for the "regular" methods and `application/json` for the "fromCustomContent" methods). In the future, we could expect that there will be other things that can't be set using the higher level APIs.

I am proposing that the PublishEvent* APIs be rationalized:

1. Ideally we would have less than 4 methods
2. It should be possible to have access to a lower-level function that allows passing the protobuf object directly or a similar object (such as [`SaveBulkState`](https://github.com/dapr/go-sdk/blob/aec809a7d5dfda7c9d5e45d81b10db17481cb1f6/client/state.go#L216) for saving state)

For example, it could be:

- Deprecate PublishEventfromCustomContentWithMetadata and remove it in version 2.0
- Change PublishEvent's last parameter to be `interface{}`:
  - If the argument is a []byte or string, it will maintain the current behavior
  - If it's a protobuf object or something similar to what SaveBulkState uses, then use that object which allows setting all options.
  - If it's something else, it will be JSON-ified like PublishEventfromCustomContentWithMetadata does today (and set content type to `application/json`)

This will remain backwards-compatible today (PublishEventfromCustomContentWithMetadata can print a log message that it's deprecated but can continue to work until v2.0).

I am happy to propose a PR if you accept this suggestion, or discuss it further.

**Comment by daixiang0** [comment]
Good catch, we can pass ContentType as param or detect in data, it would reduce duplicate functions.

**Comment by ItalyPaleAle** [comment]
@daixiang0 The thing is, today what we are missing is ContenType (and metadata). Tomorrow will be something else. What I'm proposing up there is a rationalization that, while maintaining backwards compatibility (and deprecating only 1 function), it allows for future expansion if people pass an object. If they continue passing a byte slice or string, nothing changes.

Yes perhaps i've chosen the wrong category.

**Comment by yaron2** [comment]
@ItalyPaleAle thanks for picking this up. I would agree with @daixiang0 and would like to work towards consolidating the number of functions rather than increasing them.

I'm wondering if we can use and optional parameters approach, something like

```go
PublishEvent(ctx context.Context, pubsubName, topicName string, data []byte, opts ...PubSubOption) error
```

```go
opts := []PubSubOption{
  pubsub.WithMetadata(metadata),
  pubsub.WithCustomContent(content),
}

client.PublishEvent(ctx, "pusbub", "topic", []byte{}, opts...) error
```

WDYT?

**Comment by ItalyPaleAle** [comment]
I think it would work and can look into that.

The only thing now is what do you want to do with `PublishEventfromCustomContent`? Should that still be deprecated?

**Comment by yaron2** [comment]
If we go towards the one function route, it can be deprecated.

**Comment by ItalyPaleAle** [comment]
This issue was just assigned to me, and I'm happy to continue stewarding this change. However, I am not sure what the next action item should be.

#179 should contain a fix for this issue and it's currently pending review. @yaron2 do you need me to make any change to that PR?

**Comment by yaron2** [comment]
> This issue was just assigned to me, and I'm happy to continue stewarding this change. However, I am not sure what the next action item should be.
> 
> #179 should contain a fix for this issue and it's currently pending review. @yaron2 or @pkedy do you need me to make any change to that PR?

I pinged @beiwei30 to review as they are heavy users of the Go SDK. I want to see this merged soon..
